src/keys/bip39.rs

```diff
@@ -5,24 +5,146 @@
 // "String slices are always valid UTF-8."
-type Mnemonic = str;
-type Passphrase = str;
-type Seed = [u8; 64];
 
 use alloc::string::{String, ToString};
+use core::convert::TryFrom;
+use core::ops::Deref;
 
-use unicode_normalization::UnicodeNormalization;
+use unicode_normalization::{is_nfkd, UnicodeNormalization};
+use zeroize::{Zeroize, ZeroizeOnDrop, Zeroizing};
 
-pub fn mnemonic_to_seed(m: &Mnemonic, p: &Passphrase, s: &mut Seed) {
-    let m = m.chars().nfkd().collect::<String>();
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    InvalidEntropyCount(usize),
+    NoSuchWord(String),
+    ChecksumMismatch,
+    UnnormalizedMnemonic,
+    UnnormalizedPassphrase,
+    BadWordlist,
+    BadSeparator,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct MnemonicRef<'a>(&'a str);
+
+impl<'a> Deref for MnemonicRef<'a> {
+    type Target = str;
+    fn deref(&self) -> &str {
+        self.0
+    }
+}
+
+impl<'a> TryFrom<&'a str> for MnemonicRef<'a> {
+    type Error = Error;
+    fn try_from(mnemonic_str: &'a str) -> Result<Self, Error> {
+        if is_nfkd(mnemonic_str) {
+            Ok(MnemonicRef(mnemonic_str))
+        } else {
+            Err(Error::UnnormalizedMnemonic)
+        }
+    }
+}
+
+impl<'a> PartialEq<str> for MnemonicRef<'a> {
+    fn eq(&self, other: &str) -> bool {
+        self.0.eq(other)
+    }
+}
+
+#[derive(Clone, ZeroizeOnDrop)]
+pub struct Mnemonic(String);
+
+impl Deref for Mnemonic {
+    type Target = str;
+    fn deref(&self) -> &str {
+        &self.0
+    }
+}
+
+impl From<&str> for Mnemonic {
+    fn from(unnormalized_mnemonic: &str) -> Self {
+        Self(unnormalized_mnemonic.chars().nfkd().collect())
+    }
+}
+
+impl<'a> From<&'a Mnemonic> for MnemonicRef<'a> {
+    fn from(mnemonic_ref: &'a Mnemonic) -> Self {
+        Self(&mnemonic_ref.0)
+    }
+}
+
+impl AsRef<str> for Mnemonic {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct PassphraseRef<'a>(&'a str);
+
+impl<'a> Deref for PassphraseRef<'a> {
+    type Target = str;
+    fn deref(&self) -> &str {
+        self.0
+    }
+}
 
+impl<'a> TryFrom<&'a str> for PassphraseRef<'a> {
+    type Error = Error;
+    fn try_from(passphrase_str: &'a str) -> Result<Self, Error> {
+        if is_nfkd(passphrase_str) {
+            Ok(PassphraseRef(passphrase_str))
+        } else {
+            Err(Error::UnnormalizedPassphrase)
+        }
+    }
+}
+
+#[derive(Clone, ZeroizeOnDrop)]
+pub struct Passphrase(String);
+
+impl Deref for Passphrase {
+    type Target = str;
+    fn deref(&self) -> &str {
+        &self.0
+    }
+}
+
+impl From<&str> for Passphrase {
+    fn from(unnormalized_passphrase: &str) -> Self {
+        Self(unnormalized_passphrase.chars().nfkd().collect())
+    }
+}
+
+impl<'a> From<&'a Passphrase> for PassphraseRef<'a> {
+    fn from(passphrase_ref: &'a Passphrase) -> Self {
+        Self(&passphrase_ref.0)
+    }
+}
+
+#[derive(Clone, ZeroizeOnDrop)]
```

Comment:
can probably also derive `Eq`, `PartialEq`, and `Zeroize`.

Reply:
It's a bad idea to compare secrets. At least with the default Eq. It's not constant time and may lead some information about the secret. But why would you need Eq impl?

Reply:
good counter question: it's probably quite handy for tests. Like you have maybe different ways of creating the mnemonic and you want to check whether that results in the same mnemonic.

Reply:
Tests can use `as_ref()` to extract the inner value.

Reply:
okay :+1: